Git Spice Panicking when git-spice branch submit runs on a newly tracked branch

Hello,

I haven't figured out under what situation this reproduces, but it has happened multiple times when I try to submit a newly tracked branch with not remote.

`gs b submit` -> Git spice would show me git errors saying could not set up upstream branch, then tell me a CR has been successfully created.
`gs b submit` -> On subsequent gs branch submit commands, git-spice would panic. 

```bash
╭─ ~/g/s/github.com/DataDog/web-ui evi/pawdiff/persist-viewed-items *1 +3 !3 ── 01:12:54 PM
╰─❯ gs stack restack
INF evi/pawdiff/persist-viewed-items: restacked on prod

╭─ ~/g/s/g/DataDog/web-ui evi/pawdiff/persist-viewed-items *1 +3 !3 ──────── 6s 01:13:06 PM
╰─❯ gs b submit          
Title: feat: persist viewed items 
Template: PULL_REQUEST_TEMPLATE.md
Body: Press [e] to open nvim or [enter/tab] to skip
Draft: [Y/n]
WRN Could not set upstream  branch=evi/pawdiff/persist-viewed-items remote=origin 
  error=
  | git branch: exit status 128
  | stderr:
  | fatal: the requested upstream branch 'origin/evi/pawdiff/persist-viewed-items' does not exist
  | hint:
  | hint: If you are planning on basing your work on an upstream
  | hint: branch that already exists at the remote, you may need to
  | hint: run "git fetch" to retrieve it.
  | hint:
  | hint: If you are planning to push out a new local branch that
  | hint: will track its remote counterpart, you may want to use
  | hint: "git push -u" to set the upstream config as you push.
  | hint: Disable this message with "git config set advice.setUpstreamFailure false"
INF Created #252855: https://github.com/DataDog/web-ui/pull/252855

╭─ ~/g/s/g/DataDog/web-ui evi/pawdiff/persist-viewed-items *1 +3 !3 ─────── 48s 01:14:00 PM
╰─❯ gs b submit
panic: upstream branch must be set if branch has a CR

goroutine 1 [running]:
go.abhg.dev/gs/internal/must.panicErrorf(...)
        go.abhg.dev/gs/internal/must/must.go:85
go.abhg.dev/gs/internal/must.NotBeBlankf[...]({0x0?, 0x140000e4550?}, {0x101265ac9?, 0x2e}, {0x0, 0x0, 0x0})
        go.abhg.dev/gs/internal/must/must.go:52 +0x74
go.abhg.dev/gs/internal/handler/submit.(*Handler).submitBranch(0x14000240000, {0x10155cea8, 0x140000e4550}, {0x14000234140, 0x20}, 0x140002206c0)
        go.abhg.dev/gs/internal/handler/submit/handler.go:746 +0xb3c
go.abhg.dev/gs/internal/handler/submit.(*Handler).Submit(0x14000240000, {0x10155cea8, 0x140000e4550}, 0x1400022c100)
        go.abhg.dev/gs/internal/handler/submit/handler.go:340 +0x104
main.(*branchSubmitCmd).Run(0x140001a95d8, {0x10155cea8, 0x140000e4550}, 0x1?, {0x10155a6e8, 0x14000240000})
        go.abhg.dev/gs/branch_submit.go:111 +0x148
reflect.Value.call({0x10141c480?, 0x140001a95d8?, 0x1400019d368?}, {0x10124598c, 0x4}, {0x14000218180, 0x3, 0x100ebf984?})
        reflect/value.go:581 +0x960
reflect.Value.Call({0x10141c480?, 0x140001a95d8?, 0x1400019d518?}, {0x14000218180?, 0x101543780?, 0x1013c0880?})
        reflect/value.go:365 +0x94
github.com/alecthomas/kong.callAnyFunction({0x10141c480?, 0x140001a95d8?, 0x10141ed00?}, 0x140004c14d0)
        github.com/alecthomas/kong@v1.12.1/callbacks.go:225 +0x63c
github.com/alecthomas/kong.callFunction({0x10141c480?, 0x140001a95d8?, 0x1400019d7b0?}, 0x140004c14d0)
        github.com/alecthomas/kong@v1.12.1/callbacks.go:182 +0x16c
github.com/alecthomas/kong.(*Context).RunNode(0x1400003ec80, 0x14000364870, {0x1400019de10, 0x1, 0x1})
        github.com/alecthomas/kong@v1.12.1/context.go:846 +0x40c
github.com/alecthomas/kong.(*Context).Run(0x1400003ec80, {0x1400019de10?, 0x14000020160?, 0x2?})
        github.com/alecthomas/kong@v1.12.1/context.go:875 +0x118
main.main()
        go.abhg.dev/gs/main.go:230 +0x1210
```


I'm happy to provide any additional information (whatever would help with pinpointing the source of the issue) if someone is willing to have a look at the issue.

--- a/.changes/unreleased/Fixed-20251204-211214.yaml
+++ b/.changes/unreleased/Fixed-20251204-211214.yaml
@@ -1,0 +1,3 @@
+kind: Fixed
+body: 'branch submit: Check if git configuration prevents pushed branches from being tracked, which could leave branches in a state where follow-up submits are not possible'
+time: 2025-12-04T21:12:14.799178-08:00

--- a/.changes/unreleased/Fixed-20251204-211302.yaml
+++ b/.changes/unreleased/Fixed-20251204-211302.yaml
@@ -1,0 +1,3 @@
+kind: Fixed
+body: 'branch submit: Fix panic when updating CRs that have no upstream branch configured'
+time: 2025-12-04T21:13:02.794912-08:00

--- a/internal/git/ref.go
+++ b/internal/git/ref.go
@@ -2,6 +2,7 @@ package git
 
 import (
 	"context"
+	"strings"
 
 	"go.abhg.dev/gs/internal/silog"
 )
@@ -12,6 +13,45 @@ type Refspec string
 
 func (r Refspec) String() string {
 	return string(r)
+}
+
+// Matches checks if a branch name matches a fetch refspec pattern.
+//
+// The refspec should be in the format used by git fetch, e.g.:
+//   - "refs/heads/foo/*" (pattern)
+//   - "+refs/heads/foo/*:refs/remotes/origin/foo/*" (full refspec with destination)
+//   - "refs/heads/master" (exact match)
+//
+// For fetch refspecs with a destination (containing ':'),
+// only the source side (left of ':') is used for matching.
+//
+// This implements the same algorithm as Git:
+// https://github.com/git/git/blob/f0ef5b6d9bcc258e4cbef93839d1b7465d5212b9/refspec.c#L298-L333
+func (r Refspec) Matches(ref string) bool {
+	pattern := strings.TrimPrefix(string(r), "+") // "+" prefix is optional
+
+	// For refspecs with destination, extract source side (left of ':').
+	// E.g.
+	// "+refs/heads/foo/*:refs/remotes/origin/foo/*" -> "refs/heads/foo/*"
+	pattern, _, _ = strings.Cut(pattern, ":")
+
+	prefix, suffix, ok := strings.Cut(pattern, "*")
+	if !ok {
+		// If this is not a pattern refspec (no '*'), do exact match
+		return pattern == ref
+	}
+
+	prefixLen := len(prefix)
+	suffixLen := len(suffix)
+	refLen := len(ref)
+
+	// Match if:
+	// 1. branchName starts with prefix
+	// 2. branchName is long enough for prefix + suffix
+	// 3. branchName ends with suffix
+	return refLen >= prefixLen+suffixLen &&
+		strings.HasPrefix(ref, prefix) &&
+		strings.HasSuffix(ref, suffix)
 }
 
 // SetRefRequest is a request to set a ref to a new hash.

--- a/internal/git/ref_test.go
+++ b/internal/git/ref_test.go
@@ -12,6 +12,198 @@ import (
 	"go.abhg.dev/gs/internal/text"
 )
 
+func TestRefspec_Matches(t *testing.T) {
+	tests := []struct {
+		name    string
+		refspec git.Refspec
+		ref     string
+		want    bool
+	}{
+		// Exact match cases
+		{
+			name:    "ExactMatch",
+			refspec: "refs/heads/main",
+			ref:     "refs/heads/main",
+			want:    true,
+		},
+		{
+			name:    "ExactMatchNoMatch",
+			refspec: "refs/heads/main",
+			ref:     "refs/heads/feature",
+			want:    false,
+		},
+		{
+			name:    "ExactMatchCaseSensitive",
+			refspec: "refs/heads/Main",
+			ref:     "refs/heads/main",
+			want:    false,
+		},
+
+		// Wildcard pattern cases - prefix only
+		{
+			name:    "WildcardPrefixMatch",
+			refspec: "refs/heads/*",
+			ref:     "refs/heads/feature",
+			want:    true,
+		},
+		{
+			name:    "WildcardPrefixMatchNested",
+			refspec: "refs/heads/*",
+			ref:     "refs/heads/feature/foo",
+			want:    true,
+		},
+		{
+			name:    "WildcardPrefixNoMatch",
+			refspec: "refs/heads/*",
+			ref:     "refs/tags/v1.0",
+			want:    false,
+		},
+		{
+			name:    "WildcardPrefixTooShort",
+			refspec: "refs/heads/*",
+			ref:     "refs/heads/",
+			want:    true, // empty suffix, so this matches
+		},
+
+		// Wildcard pattern cases - prefix and suffix
+		{
+			name:    "WildcardPrefixAndSuffixMatch",
+			refspec: "refs/heads/*/main",
+			ref:     "refs/heads/feature/main",
+			want:    true,
+		},
+		{
+			name:    "WildcardPrefixAndSuffixMatchLonger",
+			refspec: "refs/heads/*/main",
+			ref:     "refs/heads/team/feature/main",
+			want:    true,
+		},
+		{
+			name:    "WildcardPrefixAndSuffixNoMatch",
+			refspec: "refs/heads/*/main",
+			ref:     "refs/heads/feature/develop",
+			want:    false,
+		},
+		{
+			name:    "WildcardPrefixAndSuffixTooShort",
+			refspec: "refs/heads/*/main",
+			ref:     "refs/heads/main",
+			want:    false, // not long enough for prefix + suffix
+		},
+
+		// Wildcard at start
+		{
+			name:    "WildcardAtStartMatch",
+			refspec: "*/main",
+			ref:     "refs/heads/main",
+			want:    true,
+		},
+		{
+			name:    "WildcardAtStartNoMatch",
+			refspec: "*/main",
+			ref:     "refs/heads/feature",
+			want:    false,
+		},
+
+		// Refspec format handling - with '+' prefix
+		{
+			name:    "ForcePushPrefixExact",
+			refspec: "+refs/heads/main",
+			ref:     "refs/heads/main",
+			want:    true,
+		},
+		{
+			name:    "ForcePushPrefixWildcard",
+			refspec: "+refs/heads/*",
+			ref:     "refs/heads/feature",
+			want:    true,
+		},
+
+		// Refspec format handling - with destination
+		{
+			name:    "WithDestinationExact",
+			refspec: "refs/heads/main:refs/remotes/origin/main",
+			ref:     "refs/heads/main",
+			want:    true,
+		},
+		{
+			name:    "WithDestinationWildcard",
+			refspec: "refs/heads/*:refs/remotes/origin/*",
+			ref:     "refs/heads/feature",
+			want:    true,
+		},
+		{
+			name:    "WithDestinationNoMatch",
+			refspec: "refs/heads/main:refs/remotes/origin/main",
+			ref:     "refs/heads/feature",
+			want:    false,
+		},
+
+		// Refspec format handling - with both '+' and destination
+		{
+			name:    "ForcePushWithDestination",
+			refspec: "+refs/heads/*:refs/remotes/origin/*",
+			ref:     "refs/heads/feature",
+			want:    true,
+		},
+
+		// Real-world examples from Issue #962
+		{
+			name:    "Issue962MinimalRefspecMatch",
+			refspec: "+refs/heads/main:refs/remotes/origin/main",
+			ref:     "refs/heads/main",
+			want:    true,
+		},
+		{
+			name:    "Issue962MinimalRefspecNoMatch",
+			refspec: "+refs/heads/main:refs/remotes/origin/main",
+			ref:     "refs/heads/feature1",
+			want:    false,
+		},
+		{
+			name:    "Issue962StandardRefspec",
+			refspec: "+refs/heads/*:refs/remotes/origin/*",
+			ref:     "refs/heads/feature1",
+			want:    true,
+		},
+
+		// Edge cases
+		{
+			name:    "EmptyRefspec",
+			refspec: "",
+			ref:     "refs/heads/main",
+			want:    false,
+		},
+		{
+			name:    "EmptyRef",
+			refspec: "refs/heads/*",
+			ref:     "",
+			want:    false,
+		},
+		{
+			name:    "OnlyWildcard",
+			refspec: "*",
+			ref:     "anything",
+			want:    true,
+		},
+		{
+			name:    "OnlyWildcardWithColon",
+			refspec: "*:refs/remotes/origin/*",
+			ref:     "refs/heads/main",
+			want:    true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got := tt.refspec.Matches(tt.ref)
+			assert.Equal(t, tt.want, got,
+				"Refspec(%q).Matches(%q) = %v, want %v",
+				tt.refspec, tt.ref, got, tt.want)
+		})
+	}
+}
+
 func TestSetRef(t *testing.T) {
 	fixture, err := gittest.LoadFixtureScript([]byte(text.Dedent(`
 		as 'Test <test@example.com>'

--- a/internal/git/remote.go
+++ b/internal/git/remote.go
@@ -60,6 +60,32 @@ func (r *Repository) RemoteDefaultBranch(ctx context.Context, remote string) (st
 	return ref, nil
 }
 
+// RemoteFetchRefspecs returns the fetch refspecs for a remote.
+// That is, the refspecs used when fetching from the remote.
+// Example:
+//
+//	+refs/heads/*:refs/remotes/origin/*
+func (r *Repository) RemoteFetchRefspecs(ctx context.Context, remote string) ([]Refspec, error) {
+	output, err := r.gitCmd(
+		ctx, "config", "--get-all", "remote."+remote+".fetch").
+		OutputString(r.exec)
+	if err != nil {
+		return nil, fmt.Errorf("config get-all remote.%s.fetch: %w", remote, err)
+	}
+
+	var refspecs []Refspec
+	for line := range strings.SplitSeq(output, "\n") {
+		line = strings.TrimSpace(line)
+		if line == "" {
+			continue
+		}
+
+		refspecs = append(refspecs, Refspec(line))
+	}
+
+	return refspecs, nil
+}
+
 // RemoteRef is a reference in a remote Git repository.
 type RemoteRef struct {
 	// Name is the full name of the reference.

--- a/internal/handler/submit/handler.go
+++ b/internal/handler/submit/handler.go
@@ -8,6 +8,7 @@ import (
 	"encoding"
 	"errors"
 	"fmt"
+	"os"
 	"slices"
 	"sort"
 	"strconv"
@@ -35,6 +36,7 @@ type GitRepository interface {
 	SetBranchUpstream(ctx context.Context, branch string, upstream string) error
 	Var(ctx context.Context, name string) (string, error)
 	CommitMessageRange(ctx context.Context, start string, stop string) ([]git.CommitMessage, error)
+	RemoteFetchRefspecs(ctx context.Context, remote string) ([]git.Refspec, error)
 }
 
 var _ GitRepository = (*git.Repository)(nil)
@@ -648,6 +650,49 @@ func (h *Handler) submitBranch(
 			return status, nil
 		}
 
+		// Sanity check:
+		// If we're going to push the branch,
+		// make sure that the fetch refspec will actually fetch it.
+		// Otherwise, we will push to origin/feature,
+		// but won't have a local refs/remotes/origin/feature
+		// to track it after a 'git fetch'.
+		if refspecs, err := h.Repository.RemoteFetchRefspecs(ctx, remote); err != nil {
+			log.Warn("Unable to verify remote's fetch refspecs",
+				"remote", remote,
+				"error", err)
+		} else {
+			wantMatch := "refs/heads/" + upstreamBranch
+			var hasMatch bool
+			for _, refspec := range refspecs {
+				if refspec.Matches(wantMatch) {
+					hasMatch = true
+					break
+				}
+			}
+
+			if !hasMatch && !opts.Force {
+				log.Errorf("Remote '%v' has refspecs:", remote)
+				for _, refspec := range refspecs {
+					log.Errorf("  - %v", refspec)
+				}
+				user := cmp.Or(os.Getenv("USER"), "yourname")
+				log.Errorf("None of these will fetch branch '%v' after pushing.", upstreamBranch)
+				log.Error("This will make follow up changes on them impossible.")
+				log.Error("To fix this, you can do one of the following:")
+				log.Errorf("1. Manually add a fetch refspec for just this branch:")
+				log.Errorf("       git config --add remote.%v.fetch +refs/heads/%v:refs/remotes/%v/%v",
+					remote, upstreamBranch, remote, upstreamBranch)
+				log.Errorf("2. Prefix all your branches with your username (e.g. '%v/%v'),", user, upstreamBranch)
+				log.Errorf("   and add a fetch refspec to fetch all branches under that prefix:")
+				log.Errorf("       git config --add remote.%v.fetch '+refs/heads/%v/*:refs/remotes/%v/%v/*'",
+					remote, user, remote, user)
+				log.Errorf("   You can configure git-spice to automatically add this prefix for future branches with:")
+				log.Errorf("       git config --global spice.branchCreate.prefix %v/", user)
+				log.Errorf("3. Use the --force flag to push anyway (not recommended).")
+				return status, errors.New("remote cannot fetch pushed branch")
+			}
+		}
+
 		var prepared *preparedBranch
 		if opts.Publish {
 			needsNavComment()
@@ -752,7 +797,14 @@ func (h *Handler) submitBranch(
 		}
 	} else {
 		needsNavComment()
-		must.NotBeBlankf(upstreamBranch, "upstream branch must be set if branch has a CR")
+		if upstreamBranch == "" {
+			log.Error("No upstream branch was found for branch %v with existing CR %v", branchToSubmit, existingChange.ID)
+			log.Error("We cannot update the CR without an upstream branch name.")
+			log.Error("To fix this, identify the correct upstream branch name and set it with, e.g.:")
+			log.Error("  git branch --set-upstream-to=<remote>/<branch> %v", branchToSubmit)
+			log.Error("Then, try again.")
+			return status, errors.New("upstream branch not set")
+		}
 
 		if !opts.Publish {
 			log.Warnf("Ignoring --no-publish: %s was already published: %s", branchToSubmit, existingChange.URL)

--- a/testdata/script/issue962_branch_submit_minimal_fetch_refspec.txt
+++ b/testdata/script/issue962_branch_submit_minimal_fetch_refspec.txt
@@ -1,0 +1,111 @@
+# https://github.com/abhinav/git-spice/issues/962
+#
+# When git is configured with a minimal fetch refspec
+# that doesn't include the pushed branch,
+# the upstream tracking fails to be set after push.
+# On the second submit, git-spice panics
+# because it expects upstream to be set if a CR exists.
+
+as 'Test <test@example.com>'
+at '2024-05-18T13:57:12Z'
+
+# setup
+cd repo
+git init
+git commit --allow-empty -m 'Initial commit'
+
+# set up a fake GitHub remote
+shamhub init
+shamhub register alice
+shamhub new origin alice/example.git
+git push origin main
+
+env SHAMHUB_USERNAME=alice
+gs auth login
+
+# Configure git with a minimal fetch refspec that only fetches main.
+# This simulates the condition where pushed branches don't get fetched back.
+git config --unset-all remote.origin.fetch
+git config --add remote.origin.fetch '+refs/heads/main:refs/remotes/origin/main'
+
+# create a new branch and submit it
+git add feature1.txt
+gs bc -m 'Add feature1' feature1
+
+# Submit fails because feature1 is not covered by the fetch refspec.
+! gs branch submit --fill
+cmp stderr $WORK/golden/refspec_does_not_match.txt
+
+# Test 1: Fix with --force
+gs branch submit --fill --force
+
+# Follow-up submit fails again because upstream is still not set.
+cp $WORK/extra/feature1-update.txt feature1.txt
+git add feature1.txt
+git commit -m 'update feature1'
+! gs branch submit
+cmp stderr $WORK/golden/no_upstream_set.txt
+
+# Test 2: Fix by adding a fetch refspec
+gs trunk
+git add feature2.txt
+gs bc -m 'Add feature2' feature2
+
+# Submit fails with a similar error.
+! gs branch submit --fill
+stderr 'None of these will fetch branch'
+stderr 'To fix this, you can do one of the following:'
+stderr 'git config --add'
+
+# Fix by adding a fetch refspec for feature* branches.
+git config --add remote.origin.fetch '+refs/heads/feature*:refs/remotes/origin/feature*'
+
+# Submit succeeds now.
+gs branch submit --fill
+
+# Follow-up submit also works.
+cp $WORK/extra/feature2-update.txt feature2.txt
+git add feature2.txt
+git commit -m 'update feature2'
+gs branch submit
+
+# With the new refspec, we can also submit feature1 updates now.
+git checkout feature1
+git fetch
+git branch --set-upstream-to=origin/feature1
+gs branch submit
+
+-- repo/feature1.txt --
+Contents of feature1
+
+-- extra/feature1-update.txt --
+New contents of feature1
+
+-- repo/feature2.txt --
+Contents of feature2
+
+-- extra/feature2-update.txt --
+Updated contents of feature2
+
+-- golden/refspec_does_not_match.txt --
+ERR Remote 'origin' has refspecs:
+ERR   - +refs/heads/main:refs/remotes/origin/main
+ERR None of these will fetch branch 'feature1' after pushing.
+ERR This will make follow up changes on them impossible.
+ERR To fix this, you can do one of the following:
+ERR 1. Manually add a fetch refspec for just this branch:
+ERR        git config --add remote.origin.fetch +refs/heads/feature1:refs/remotes/origin/feature1
+ERR 2. Prefix all your branches with your username (e.g. 'yourname/feature1'),
+ERR    and add a fetch refspec to fetch all branches under that prefix:
+ERR        git config --add remote.origin.fetch '+refs/heads/yourname/*:refs/remotes/origin/yourname/*'
+ERR    You can configure git-spice to automatically add this prefix for future branches with:
+ERR        git config --global spice.branchCreate.prefix yourname/
+ERR 3. Use the --force flag to push anyway (not recommended).
+FTL gs: submit branch feature1: remote cannot fetch pushed branch
+-- golden/no_upstream_set.txt --
+ERR No upstream branch was found for branch %v with existing CR %v  feature1=#1
+ERR We cannot update the CR without an upstream branch name.
+ERR To fix this, identify the correct upstream branch name and set it with, e.g.:
+ERR   git branch --set-upstream-to=<remote>/<branch> %v  !BADKEY=feature1
+ERR Then, try again.
+FTL gs: submit branch feature1: upstream branch not set

